packages/web/components/drawers/token-select-drawer-limit.tsx

```diff
@@ -0,0 +1,445 @@
+    const [confirmUnverifiedAssetDenom, setConfirmUnverifiedAssetDenom] =
+      useState<string | null>(null);
+
+    const showUnverifiedAssetsSetting =
+      userSettings.getUserSettingById<UnverifiedAssetsState>(
+        "unverified-assets"
+      );
+    const shouldShowUnverifiedAssets =
+      showUnverifiedAssetsSetting?.state.showUnverifiedAssets;
+
+    const assetsRef = useLatest(assets);
+
+    const searchBoxRef = useRef<HTMLInputElement>(null);
+    const quickSelectRef = useRef<HTMLDivElement>(null);
+
+    const { onMouseDown: onMouseDownQuickSelect } =
+      useDraggableScroll(quickSelectRef);
+
+    const onClose = () => {
+      setIsRequestingClose(true);
+      setKeyboardSelectedIndex(0);
+      onCloseProp?.();
+    };
+
+    const onSelect = (coinDenom: string) => {
+      onSelectProp?.(coinDenom);
+      onClose();
+    };
+
+    const onClickAsset = (coinDenom: string) => {
+      let isRecommended = false;
+      const selectedAsset = assets.find(
+        (asset) => asset.coinDenom === coinDenom
+      );
+
+      // shouldn't happen, but doing nothing is better
+      if (!selectedAsset) return;
+
+      if (
+        !isRecommended &&
+        !shouldShowUnverifiedAssets &&
+        !selectedAsset.isVerified
+      ) {
+        return setConfirmUnverifiedAssetDenom(coinDenom);
+      }
+
+      onSelect(coinDenom);
+    };
+
+    useWindowKeyActions({
+      Escape: onClose,
+    });
+
+    const { handleKeyDown: containerKeyDown } = useKeyActions({
+      ArrowDown: () => {
+        setKeyboardSelectedIndex((selectedIndex) =>
+          selectedIndex === getAllTokenElements().length - 1
+            ? 0
+            : selectedIndex + 1
+        );
+
+        getTokenElement(
+          uniqueId,
+          keyboardSelectedIndexRef.current
+        )?.scrollIntoView({
+          block: "nearest",
+        });
+
+        // Focus on search bar if user starts keyboard navigation
+        searchBoxRef.current?.focus();
+      },
+      ArrowUp: () => {
+        setKeyboardSelectedIndex((selectedIndex) =>
+          selectedIndex === 0
+            ? getAllTokenElements().length - 1
+            : selectedIndex - 1
+        );
+
+        getTokenElement(
+          uniqueId,
+          keyboardSelectedIndexRef.current
+        )?.scrollIntoView({
+          block: "nearest",
+        });
+
+        // Focus on search bar if user starts keyboard navigation
+        searchBoxRef.current?.focus();
+      },
+      Enter: () => {
+        const asset = assetsRef.current[keyboardSelectedIndexRef.current];
+        if (!asset) return;
+        const { coinDenom } = asset;
+
+        onClickAsset(coinDenom);
+      },
+    });
+
+    const { handleKeyDown: searchBarKeyDown } = useKeyActions({
+      ArrowDown: (event) => {
+        event.preventDefault();
+      },
+      ArrowUp: (event) => {
+        event.preventDefault();
+      },
+    });
+
+    // TODO: Wire up search
```

Comment:
We have a `useFilteredData` hook that can do fuzzy search on client.

Reply:
https://github.com/osmosis-labs/osmosis-frontend/pull/3308/commits/1a3e662ad0263660bb65f96c479f75a1c6a1df35